File: Algorand-NFT-SDK/python-algorand-sdk/py_sdk/testnet.py
import hashlib
import json
from algosdk.v2client import algod, indexer
from algosdk import transaction, account, mnemonic

algod_address_testnet = "https://testnet-api.algonode.cloud"
indexer_address_testnet = "https://testnet-idx.algonode.cloud"
algod_token = ""
indexer_token = ""
admin_address = "W4BERQ52RZILAKXNJJ6X5FNY3ASIAK3OV6KWX7DRTLKHXE7HNNGCO5OVUA"
admin_key = "3d43u0uywHWdackLRDwgZA8cBqpb2FxUJECl36fhcvS3AkjDuo5QsCrtSn1+lbjYJIArbq+Va/xxmtR7k+drTA=="

algod_client = algod.AlgodClient(algod_token, algod_address_testnet)
indexer_client = indexer.IndexerClient(indexer_token, indexer_address_testnet)

def create_nft(name, symbol, metadata_url, reserve="", freeze="", clawback=""):
  """
  returns the id of the created NFT
  """
  metadata_hash = hashlib.new("sha256", metadata_url.encode())
  metadata_bytes = metadata_hash.digest()
  
  print("Your metadata hash: {}".format(metadata_hash.hexdigest()))

  # build unsigned transaction

  sp = algod_client.suggested_params()

  txn = transaction.AssetConfigTxn(
    sender=admin_address,
    sp=sp,
    total=1,
    decimals=0,
    default_frozen=False,
    unit_name=symbol,
    asset_name=name,
    manager=admin_address,
    strict_empty_address_check=False,
    reserve=reserve,
    freeze=freeze,
    clawback=clawback,
    url=metadata_url,
    metadata_hash=metadata_bytes 
  )

  # sign transaction
  signed_txn = txn.sign(admin_key)

  #submit transaction
  txid = algod_client.send_transaction(signed_txn)

  print("Successfully sent transaction with txID: {}".format(txid))

  # wait for confirmation 
  try:
    confirmed_txn = transaction.wait_for_confirmation(algod_client, txid, 4)  
  except Exception as err:
    print(err)
    return False

  print("Transaction information: {}".format(json.dumps(confirmed_txn, indent=4)))

  print("Asset ID: {}".format(confirmed_txn["asset-index"]))

  return confirmed_txn["asset-index"]
 
def verify_opt_in(address, asset_id):
  account_info = indexer_client.lookup_account_assets(address=address, asset_id=asset_id)
  
  if account_info.get("assets")[0].get("asset-id") == asset_id:
    return True
  else:
    return False
 
def claim_nft(address, asset_id):
  
  if (verify_opt_in(address=address, asset_id=asset_id)) == False:
    print("Asset with Id: {} has not yet been opted into by Wallet: {}. Please opt-in and try again").format(asset_id, address)
    return False
   
  sp = algod_client.suggested_params()
 
  txn = transaction.AssetTransferTxn(
    sender=admin_address,
    sp=sp,
    receiver=address,
    amt=1,
    index=asset_id  
  )
 
  signed_txn = txn.sign(admin_key)

  txid = algod_client.send_transaction(signed_txn)
  print("Successfully sent transaction with txID: {}".format(txid))

  try:
    confirmed_txn = transaction.wait_for_confirmation(algod_client, txid, 4)

  except Exception as err:
    print(err)
    return False

  print("Transaction information: {}".format(json.dumps(confirmed_txn, indent=4)))

  return txid
 
def create_account():
  private_key, address = account.generate_account()
  print("My address: {}".format(address))
  print("My private key: {}".format(private_key))
  print("My passphrase: {}".format(mnemonic.from_private_key(private_key)))
  
  return address
 
def update_nft(asset_id, reserve="", clawback="", freeze=""):
  sp = algod_client.suggested_params()
   
  txn = transaction.AssetConfigTxn(
    sender=admin_address,
    sp=sp,
    default_frozen=False,
    index=asset_id,
    manager=admin_address,
    reserve=reserve,
    freeze=freeze,
    clawback=clawback,    
  )
  
    # sign transaction
  signed_txn = txn.sign(admin_key)

  #submit transaction
  txid = algod_client.send_transaction(signed_txn)

  print("Successfully sent transaction with txID: {}".format(txid))

  # wait for confirmation 
  try:
    confirmed_txn = transaction.wait_for_confirmation(algod_client, txid, 4)  
  except Exception as err:
    print(err)
    return False

  print("Transaction information: {}".format(json.dumps(confirmed_txn, indent=4)))

  print("Asset ID: {}".format(confirmed_txn["asset-index"]))

  return confirmed_txn["asset-index"]
 
 
 
 

File: Algorand-NFT-SDK/javaScript-sdk/testnet.js
const { generateAccount } = require('algosdk')
const algosdk = require('algosdk')
const crypto = require('crypto')
const fs = require('fs')
const { createInflate } = require('zlib')

const algodTestnetServer = 'https://testnet-api.algonode.cloud'
const indexerTestnetServer = 'https://testnet-api.algonode.cloud'
const algodBetanetServer = 'https://betanet-api.algonode.cloud'
const indexerBetanetServer = 'https://betanet-idx.algonode.cloud'
const algodMainetServer = 'https://mainnet-api.algonode.cloud'
const indexerMainnetServer = 'https://mainnet-idx.algonode.cloud'
const algodPort = 443
const indexerPort = 443
const algodToken = ''
const indexerToken = ''

const adminAddress =
	'ONOL67X6NTUZSMWUQPNHZRFXCO7JK547YYJCSDQTI5V3BOFPPJRK73EFNA'
const mnemonic =
	'pink sound impulse project salt asset until leg glide kite inhale salt logic sell pull street gun viable outdoor electric funny club journey above estate'

const recoveredAccount = algosdk.mnemonicToSecretKey(mnemonic);

const adminKey = recoveredAccount.sk;

const algodClient = new algosdk.Algodv2(
	algodToken,
	algodTestnetServer,
	algodPort
)
const indexerClient = new algosdk.Indexer(
	indexerToken,
	indexerTestnetServer,
	indexerPort
)

const createAccount = function () {
	try {
		const myaccount = algosdk.generateAccount()
		console.log('Account Address = ' + myaccount.addr)
		let account_mnemonic = algosdk.secretKeyToMnemonic(myaccount.sk)
		console.log('Account Mnemonic = ' + account_mnemonic)
		console.log('Account created. Please store Mnemonic and address')
		console.log('Add funds to account using the TestNet Dispenser: ')
		console.log(
			'https://dispenser.testnet.aws.algodev.network/?account=' + myaccount.addr
		)

		return myaccount
	} catch (err) {
		console.log({ err })
	}
}

async function checkOptIn(address, assetId) {
	const accountInfo = await algodClient.accountInformation(address).do()
	const assets = accountInfo.assets
	// console.log({assets});

	let optInStatus = false
	for (const asset of assets) {
		// console.log({asset})
		if (asset['asset-id'] === parseInt(assetId)) {
			console.log(asset['asset-id'])
			optInStatus = true
			break
		}
	}
	return optInStatus
}

async function claimNFT(address, assetId) {
	const isOptIn = await checkOptIn(address, assetId)
	if (!isOptIn) {
		console.log(
			`Asset with Id: ${assetId} has not yet been opted into by Wallet: ${address}. Please opt-in`
		)
		return false
	}

	const sp = await algodClient.getTransactionParams().do()
	const txn = algosdk.makeAssetTransferTxnWithSuggestedParamsFromObject({
		amount: 1,
		assetIndex: assetId,
		closeRemainderTo: undefined,
		from: adminAddress,
		note: undefined,
		rekeyTo: undefined,
		revocationTarget: undefined,
		suggestedParams: sp,
		to: address,
 })
	const signedTxn = txn.signTxn(adminKey)
	const tx = await algodClient.sendRawTransaction(signedTxn).do()
	console.log(`Successfully sent transaction with txID: ${tx.txId}`)

	const confirmedTxn = await algosdk.waitForConfirmation(
		algodClient,
		tx.txId,
		10
	)
	console.log(
		`Transaction information: ${JSON.stringify(confirmedTxn, null, 4)}`
	)
	console.log(`Transaction ID: ${tx.txId}`)

	return tx.txId
}

const verifyOptInBeforeTransfer = async (address, asset) => {
	new Promise((resolve) => {
		let waiter = undefined
		waiter = setInterval(async () => {
			const hasOptIn = await checkOptIn(address, asset) // true || false
			if (hasOptIn) {
				clearInterval(waiter)
				waiter = undefined
				resolve()
			}
		}, 5000)
	})
		.then(async () => {
			await claimNFT(address, asset)
		})
		.catch((error) => {
			console.log({ error })
		})
}

async function createNft({
	name = undefined,
	symbol = undefined,
	address = undefined,
	url = undefined,
	clawback = undefined,
	freeze = undefined,
} = {}) {
	if (name === undefined || symbol === undefined || url === undefined) {
		console.log('Argument missing required parameter')
		return
	}

	if (address === undefined) {
		generateAccount()
	}

	const sp = await algodClient.getTransactionParams().do()

	// todo create if statement to check if reserve and if reserve, build metadata hash else empty string
	const hash = crypto.createHash('sha256')
	hash.update(url)
	const digest = hash.digest()
	const hexDigest = digest.toString('hex')
	console.log(`Your metadata hash: ${hexDigest}`)

	const stringToUint8Array = (str) => {

		const utf8 = decodeURI(encodeURIComponent(str))

		const array = new Uint8Array(utf8.split('').map((item) => {

			return item.charCodeAt(0)

	}))

		return array

	}

	// function b64_to_utf8(str) { return decodeURIComponent((atob(str))); }
	
	const txn = algosdk.makeAssetCreateTxnWithSuggestedParamsFromObject({
		// todo may change url p
		assetMetadataHash: stringToUint8Array(""),
		assetName: name,
		assetURL: url,
		clawback: clawback,
		decimals: 0,
		defaultFrozen: false,
		freeze: freeze,
		from: adminAddress,
		manager: adminAddress,
		reserve: undefined,
		suggestedParams: sp,
		total: 1,
		unitName: symbol,
		strictEmptyAddressChecking: false,
	})

	const rawSignedTxn = txn.signTxn(adminKey)

	let tx = await algodClient.sendRawTransaction(rawSignedTxn).do()
	console.log('Transaction : ' + tx.txId)
	let assetID = null
	// wait for transaction to be confirmed

	try {
		const confirmedTxn = await algosdk.waitForConfirmation(
			algodClient,
			tx.txId,
			10
		)

		// Get the new asset's information from the creator account
		let ptx = await algodClient.pendingTransactionInformation(tx.txId).do()
		assetID = ptx['asset-index']

		console.log(
			`Transaction information: ${JSON.stringify(confirmedTxn, null, 4)}`
		)
		console.log(`Asset ID: ${confirmedTxn["asset-index"]}`)
		console.log(`Please copy Asset ID and opt-in to the asset in your wallet.`)
	} catch (err) {
		console.error({ err })
		return false
	}
	verifyOptInBeforeTransfer(address, assetID)
}

// todo update nft function
async function updateNFT({
	assetId = undefined,
	reserve = undefined,
	clawback = undefined,
	freeze = undefined,
} = {}) {
	if (assetId === undefined || reserve === undefined) {
		console.log('Argument missing required parameter')
		return
	}

	const suggestedParams = await algosdk.algod.getTransactionParams()

	const txn = algosdk.makeAssetConfigTransaction(
		adminAddress,
		{
			index: assetId,
			manager: adminAddress,
			reserve: reserve,
			freeze: freeze,
			clawback: clawback,
			defaultFrozen: false,
			strictEmptyAddressChecking: false,
		},
		suggestedParams
	)

	const rawSignedTxn = txn.signTxn(adminKey)

	let tx = await algodClient.sendRawTransaction(rawSignedTxn).do()
	console.log('Transaction : ' + tx.txId)
	let assetID = null
	// wait for transaction to be confirmed

	try {
		const confirmedTxn = await algosdk.waitForConfirmation(
			algodClient,
			tx.txId,
			10
		)

		// Get the new asset's information from the creator account
		let ptx = await algodClient.pendingTransactionInformation(tx.txId).do()
		assetID = ptx['asset-index']

		console.log(
			`Transaction information: ${JSON.stringify(confirmedTxn, null, 4)}`
		)
		console.log(`Asset ID: ${confirmedTxn.txresults.createdasset}`)
	} catch (err) {
		console.error(err)
		return false
	}
}

